File: ts/256/240.ts
// 二分查找
function searchMatrix(matrix: number[][], target: number): boolean {
  const m = matrix.length, n = matrix[0].length
  const bs = (nums: number[], target: number): boolean => {
    let low = 0, high = nums.length - 1
    while (low <= high) {
      const mid = Math.floor((high - low) / 2) + low
      if (nums[mid] === target) return true
      if (nums[mid] > target) {
        high = mid - 1
      } else {
        low = mid + 1
      }
    }
    return false
  }
  for (let i = 0; i < m; i++) {
    if (bs(matrix[i], target)) return true
  }
  return false
};

// Z 字形查找
function searchMatrix1(matrix: number[][], target: number): boolean {
  const m = matrix.length, n = matrix[0].length
  let i = 0, j = n - 1
  while (i < m && j >= 0) {
    if (matrix[i][j] === target) return true
    matrix[i][j] > target ? j -= 1 : i += 1
  }
  return false
};

File: py3/256/240.py
from typing import List

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0]) if m != 0 else 0
        if n == 0:
            return False
        i, j = 0, n - 1
        while i < m and j < n:
            if matrix[i][j] == target:
                return True
            if matrix[i][j] < target:
                i += 1
            else:
                j -= 1
            if i < 0 or j < 0:
                break
        return False